Fixes bug 1306891 - Integrate new correlations results in Socorro

marco-c opened the pull request:
I'd like to get some feedback before continuing.

Don't mind the usage of ES6 features, I'll remove them.

marco-c commented:
@adngdb ping, I'd just like some quick feedback here, not a full review.

adngdb commented:
I meant to do this today and got distracted, sorry. I'll get to it tomorrow!

adngdb commented on webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js:
```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+            return sha1(signature)
+            .then(sha1signature => {
+              return sha1signature;
+            })
```
This ``then`` does nothing, right?

marco-c replied:
Yes, I was probably debugging something here :)

adngdb commented on webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js:
```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+            .then(response => response.json())
+            .then(data => {
+                correlationData[product][channel]['signatures'][signature] = data;
+            });
```
I think every Promise chain should end with, at least, ``.catch(console.log.bind(console));`` or any way to handle errors. Logging is the simplest way of handling them but I would be satisfied with it for a start.

marco-c replied:
IIRC they are logged anyway in the console if you don't catch them, but I'll add the `catch` anyway.

adngdb commented on webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js:
```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+          return '100.0';
+        }
+
+        if (result.substring(0, result.indexOf('.')).length == 1) {
```
``result`` being a number, that should raise a ``TypeError`` because (number).substring is not a function.

marco-c replied:
`result` is a string (generated with `toFixed`)

adngdb commented on webapp-django/crashstats/signature/static/signature/js/signature_tab_correlations.js:
```diff
@@ -182,6 +182,18 @@ SignatureReport.CorrelationsTab.prototype.onAjaxSuccess = function () {
+    function load() {
+        var channel = $('#correlations_channel').val();
+        var product = $('#correlations_product').val();
+        var signature = $('#correlations_signature').val();
```
You don't need to use a hidden input, the signature is already accessible with ``SignatureReport.signature``.

adngdb commented on webapp-django/crashstats/signature/views.py:
```diff
@@ -372,6 +372,21 @@ def signature_correlations(request, params):
+    if 'product' in params and 'FennecAndroid' in params['product'] and 'Firefox' not in params['product']:
+            context['product'] = 'FennecAndroid'
+
+    context['signature'] = signature
```
And thus you can remove this line as well.

peterbe commented on webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js:
```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+        }
+
+        if (result.substring(0, result.indexOf('.')).length == 1) {
+          return '0' + result;
```
This is all odd. First of all, the if statement can simply be done with `if (num < 0.01)`
but why display the number with a leading padded 0? If there's a good reason, there needs to be a comment somewhere here. 

Also, I find painful to use 2 decimals for all numbers except for `100.0`.
If the purpose of all of this is to make sure the string is always 5 characters, I'd much much rather solve the display part of this using CSS.

peterbe commented on webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js:
```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+    return [ci[0] - correction, ci[1] + correction];
+  }
+
+  function sortCorrelationData(correlationData, total_reference, total_group) {
```
I can't skim this function and understand it. It sorts number *number* of keys in each objects' `item` attribute??? And then if equal it sorts by some sort of comparison of confident intervals.

marco-c replied:
I'll add some comments.

It sorts the results by number of attributes first (e.g. platform=Windows will come before platform=Windows && graphic_card=NVIDIA), because they are often the most interesting results.

Then, it sorts by percentage difference (e.g. if platform=Windows is 100% in the signature and 60% overall and graph_card=NVIDIA is 40% in the signature and 25% overall, platform=Windows will come first).

peterbe commented on webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js:
```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+
+  window.correlations = {};
+  window.correlations.loadCorrelationData = loadCorrelationData;
+  window.correlations.writeResults = text;
```
Why these three lines? I don't get it. 
And perhaps it wasn't planned. 

So you're defining all of these functions AFTER the DOM has been loaded. Why can't the functions be define BEFORE or INDEPENDENT of the loading of the DOM?

Use the module pattern or just encapsulate all of these functions as methods in a class.

marco-c replied:
I put the code in there just temporarily, for testing.

peterbe commented on webapp-django/crashstats/signature/jinja2/signature/signature_correlations.html:
```diff
@@ -1,5 +1,29 @@
 
+    <label for="correlations_product">Product:</label>
+    <select id="correlations_product">
+        {% for p in ['Firefox', 'FennecAndroid'] %}
```
Why hardcode these?

marco-c replied:
For now, the correlations are only generated for these two products.

marco-c commented on webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js:
```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+        .then(hash => hex(hash));
+    }
+
+    let correlationData = {};
```
I'm caching the results in this variable, so they don't have to be requested multiple times (e.g. if in the correlations tab the user switches to another product and then switches back).

marco-c commented:
> So, in conclusion test by running ./manage.py collectstatic --noinput then run with DEBUG=False in your .env and see if things still work.

I've had to remove arrow functions, `let` and `for ... of`.

marco-c commented:
In another PR, I'll add the correlations results to the top crashers page, replacing the current ones.

adngdb commented:
@marco-c please don't rebase your commits before reviewing is finished, it makes my job a lot harder! Now I can't see what you have changed since the last time I reviewed. :(

Also, I thought you wanted to keep the old correlations around because modules and other stuff where missing from yours? I don't mind that you remove the current correlations, they are doomed to die anyway, but I guess I'm just confused.

marco-c commented:
> @marco-c please don't rebase your commits before reviewing is finished, it makes my job a lot harder! Now I can't see what you have changed since the last time I reviewed. :(

Oops, sorry. I thought at the beginning you only had a quick look. In general, most of the code is the same, I've just incorporated your feedback and Peter's, and I've cleaned up the old correlation stuff.

> Also, I thought you wanted to keep the old correlations around because modules and other stuff where missing from yours? I don't mind that you remove the current correlations, they are doomed to die anyway, but I guess I'm just confused. 

Talking with Peter, we concluded that we could remove the old correlation stuff because basically nobody is using it, it doesn't work most of the time, and the reports are still accessible anyway at crash-analysis.mozilla.org. Only the modules are missing for now, but hopefully they will be included soon.

adngdb commented:
I just did quite big review written down as code: https://github.com/marco-c/socorro/pull/1 :)

adngdb commented:
@marco-c Do you think this is ready to go? I haven't tested them but your last changes look good!

marco-c commented:
Yes, I think it's ready.
I've tested some signature pages, a report page for a signature with correlations, and a report page for a signature without correlations. All of them seem to be working correctly.

adngdb commented:
This is the only concern I have after testing this new feature: https://bugzilla.mozilla.org/show_bug.cgi?id=1315109

It can be a follow-up improvement. These new correlations rock! Let's ship them! 

Thanks for your impressive work @marco-c !!!

marco-c commented:
> This is the only concern I have after testing this new feature: https://bugzilla.mozilla.org/show_bug.cgi?id=1315109

Makes perfect sense. Sometimes signatures might have the same causes in Thunderbird and Firefox, so checking the correlations with the Firefox product might be useful, but it's better to avoid confusion.

Thank you for the reviews and the fixes!
